Userland/Utilities/strings.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CharacterTypes.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/Stream.h>
+#include <LibCore/System.h>
+#include <LibMain/Main.h>
+#include <unistd.h>
+
+enum class StringOffsetFormat {
+    None = 0,
+    Decimal,
+    Octal,
+    Hexadecimal
+};
+
+static bool should_print_characters(Vector<u8> const& characters)
+{
+    for (u8 ch : characters) {
+        if (is_ascii_printable(ch) && !is_ascii_space(ch))
+            return true;
+    }
+    return false;
+}
+
+static void print_characters(Vector<u8> const& characters, StringOffsetFormat string_offset_format, size_t string_offset_position)
+{
+    switch (string_offset_format) {
+    case StringOffsetFormat::Decimal:
+        outln("offset {:d}: ", string_offset_position);
+        break;
+    case StringOffsetFormat::Octal:
+        outln("offset 0o{:o}: ", string_offset_position);
+        break;
+    case StringOffsetFormat::Hexadecimal:
+        outln("offset 0x{:x}: ", string_offset_position);
+        break;
+    default:
+        break;
+    }
+    outln("{:s}", characters.span());
+}
+
+static int process_characters_in_span(Vector<u8>& characters, AK::Span<u8> const& span)
+{
+    int processed_characters = 0;
+    for (u8 ch : span) {
+        ++processed_characters;
+        if (is_ascii_printable(ch) || ch == '\t')
+            characters.append(ch);
+        else
+            break;
+    }
+    return processed_characters;
+}
+
+static ErrorOr<void> process_strings_in_file(StringView path, size_t minimum_string_length, bool show_paths, StringOffsetFormat string_offset_format)
```

Comment:
`minimum_string_length`, `show_paths` and `string_offset_format` don't change anywhere after they have been initialized and parsed (certainly not per-file), so we could probably make those global instead of passing them around as function arguments.

Reply:
> we could probably make those global instead of passing them around as function arguments

Would that not be in violation of I.2 of C++ Core Guidelines? I'm trying my best not to anger the CI deities :smile:

Reply:
As long as it's not the "Serenity C++ Core Guidelines". :^)

Looking at what the guidelines say though, I don't think that this is the case that they were thinking of.

If we use global variables to store settings for some command line tool, that is (at least in my opinion) essentially a global constant that only gets initialized once and might get updated once if the user overrides the default setting. I don't see any "hidden dependencies" or data races there.

Reply:
Ok, I have changed them to global, and CI seems to be happy with that :+1: